### DISCUSSION TITLE
Update does not enforce type on `data`

### DISCUSSION THREAD

**Discussion by GuiSim** [prisma/prisma#9957] [Q&A] [answered]
Hello! 

I'm new to Prisma and testing it out. I encountered an issue today where an `update` statement would not raise any type issue on an invalid `data` object.

```ts
prisma.mything.update({
  where: { id: '1' },
  data: { invalidField: 42 }, // Does not compile as `invalidField` is not a valid field on `mything`
});

const data = { invalidField: 42 };
prisma.mything.update({
  where: { id: '1' },
  data: data, // Compiles without issues. Why?
});
```

Can you help me understand why?

**Comment by austincrim** [comment, marked as answer]
Hey @GuiSim 👋!

This is definitely unexpected. I'm going to convert this to an issue where our engineering teams will triage and investigate further. Feel free to follow it as we might ask for more information.

We really appreciate you pointing this one out 👊.

**Reply by GuiSim** [reply]
Ah! That's so great to read. Glad I was confused by a "bug".
I understand this is more a limitation of TypeScript but it seems that there are possible workarounds.

In my humble opinion I would expect the same validations to be performed no matter how `data` is defined so if that's possible that would be a great win.

I'll follow #9988 closely! Thank you again.

**Comment by aqrln** [comment]
It's the default behavior of TypeScript and how its type system works. TypeScript checks that the object type strictly matches the expected type when you pass an object literal, but allows any compatible type with extra fields (basically, a subtype) when you pass an identifier.

It's possible to override this behavior and always match the type strictly, though, using some type system hacks like this:

```ts
export type StrictPropertyCheck<T, TExpected, TError> = T extends TExpected
  ? Exclude<keyof T, keyof TExpected> extends never
    ? T
    : TError
  : TExpected
```

Could you please open an issue?

**Reply by aqrln** [reply]
Just saw that @austincrim was going to open an issue, so I guess that's already done.

**Reply by GuiSim** [reply]
Thank you @aqrln for the technical explanation as to _why_.
It's a bit surprising to me that object literals provide more safety when passed directly as parameter than when passed through a variable.

**Reply by aqrln** [reply]
@GuiSim I would say in most cases TypeScript behavior is reasonable, it wouldn't have been an expressive enough language without subtyping and interfaces as an abstraction mechanism. For immediate object literals it is safe for TypeScript to assume that any extra fields are accidental or wrong, but in general case this would be too restrictive and would forbid a lot of useful and type-safe patterns, especially for code written in object-oriented style. Exact objects would need to be a different kind of types with a different syntax, like in Flow.

**Reply by aqrln** [reply]
Or maybe something like `StrictPropertyCheck` type from the answer above could have been a part of TypeScript or its standard library, but I don't have as much context as TypeScript team does to assess if that's a good idea.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
